File: rust_kernel/src/drivers/storage/vfs/statvfs.rs
/// This is the layout of the statvfs structure as defined by POSIX-2018.
/// unsigned long f_bsize    File system block size.
/// unsigned long f_frsize   Fundamental file system block size.
/// fsblkcnt_t    f_blocks   Total number of blocks on file system in units of f_frsize.
/// fsblkcnt_t    f_bfree    Total number of free blocks.
/// fsblkcnt_t    f_bavail   Number of free blocks available to
///                          non-privileged process.
/// fsfilcnt_t    f_files    Total number of file serial numbers.
/// fsfilcnt_t    f_ffree    Total number of free file serial numbers.
/// fsfilcnt_t    f_favail   Number of file serial numbers available to
///                          non-privileged process.
/// unsigned long f_fsid     File system ID.
/// unsigned long f_flag     Bit mask of f_flag values.
/// unsigned long f_namemax  Maximum filename length.

type FsBlockCount = usize;
type FsFileCount = usize;
type FileSystemId = usize;


#[allow(snake_case)]
enum f_flag {
    ST_RDONLY,
    ST_NOSUID,
}

struct StatVfs {
    f_bsize: usize,
    f_frsize: usize,
    f_blocks: FsBlockCount,
    f_bfree: FsBlockCount,
    f_bavail: FsBlockCount,
    f_files: FsFileCount,
    f_ffree: FsFileCount,
    f_favail: FsFileCount,
    f_fsid: FileSystemId,
    f_flag: f_flag,
    f_namemax: usize,
}

File: libc/src/sys/statvfs/statvfs.c
#include <sys/statfs.h>
#include <sys/statvfs.h>

int statvfs(const char *restrict path, struct statvfs *restrict buf)
{
	struct statfs	statfs_buf;

	int ret = statfs(path, &statfs_buf);

	if (-1 == ret) {
		return -1;
	}

	buf->f_bsize   = statfs_buf.f_bsize;
	buf->f_frsize  = statfs_buf.f_frsize;
	buf->f_blocks  = statfs_buf.f_blocks;
	buf->f_bfree   = statfs_buf.f_bfree;
	buf->f_bavail  = statfs_buf.f_bavail;
	buf->f_files   = statfs_buf.f_files;
	buf->f_ffree   = statfs_buf.f_ffree;
	buf->f_favail  = statfs_buf.f_ffree; // Supposedly what we want.
	buf->f_fsid    = statfs_buf.f_fsid;
	buf->f_flag    = statfs_buf.f_flags;
	buf->f_namemax = statfs_buf.f_namelen;
	return 0;
}
